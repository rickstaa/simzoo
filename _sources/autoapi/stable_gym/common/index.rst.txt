stable_gym.common
=================

.. py:module:: stable_gym.common

.. autoapi-nested-parse::

   Contains several functions that can be used across all the Stable Gym gymnasium
   environments.



Submodules
----------

.. toctree::
   :maxdepth: 1

   /autoapi/stable_gym/common/max_episode_steps_injection_wrapper/index
   /autoapi/stable_gym/common/utils/index


Classes
-------

.. autoapisummary::

   stable_gym.common.MaxEpisodeStepsInjectionWrapper


Package Contents
----------------

.. py:class:: MaxEpisodeStepsInjectionWrapper(env)

   Bases: :py:obj:`gymnasium.Wrapper`


   A gymnasium wrapper that injects the ``max_episode_steps`` attribute of the
   ``TimeLimit`` wrapper into the base environment as the
   ``_time_limit_max_episode_steps`` attribute. If the environment is not wrapped in
   a ``TimeLimit`` wrapper, then the ``_time_limit_max_episode_steps`` attribute is
   set to ``None``.

   Wrap a gymnasium environment.
   :param env: The gymnasium environment.
   :type env: gym.Env


   .. py:attribute:: max_episode_steps


