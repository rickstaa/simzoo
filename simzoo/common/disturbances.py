"""Common disturbances used in the disturber.
"""

import numpy as np


def impulse_disturbance(
    input_signal, impulse_magnitude, impulse_instant, impulse_type, current_timestep,
):
    """Retrieves a impulse disturbance that acts in the opposite direction of the input
    signal.

    Args:
        input_signal (numpy.ndarray): The signal to which the disturbance should be
            applied. Used for determining the direction of the disturbance.
        impulse_magnitude (float): The magnitude of the impulse disturbance.
        impulse_instant (float): The time step at which you want to apply the impulse
            disturbance.
        impulse_type (str): The type of impulse disturbance you want to use. Options
            are: 'regular' and 'constant'. A regular (instant) impulse (applied at a
            single time step) and a constant impulse (applied at all steps following the
            set time instant)
        current_timestep (int): The current time step.

    Returns:
        numpy.ndarray: The disturbance array.
    """
    if impulse_type.lower() == "constant":
        if current_timestep >= impulse_instant:
            dist_val = impulse_magnitude * (-np.sign(input_signal))
        else:
            dist_val = np.zeros_like(input_signal)
    else:
        if current_timestep == impulse_instant:
            dist_val = impulse_magnitude * (-np.sign(input_signal))
        else:
            dist_val = np.zeros_like(input_signal)
    return dist_val


def periodic_disturbance(
    input_signal, current_timestep, amplitude=1, frequency=10, phase_shift=0
):
    """Returns a periodic disturbance signal that has the same shape as the input signal.

    Args:
        input_signal (numpy.ndarray): The signal to which the disturbance should be
            applied. Used for determining the direction of the disturbance.
        current_timestep(int): The current time step.
        amplitude (float, optional): The periodic signal amplitude. Defaults to ``1``.
        frequency (float, optional): The periodic signal frequency. Defaults to ``10``.
        phase_shift (float, optional): The periodic signal phase shift. Defaults to
            ``0``.

    Returns:
        numpy.ndarray: The disturbance array.
    """
    return (
        amplitude
        * np.sin(2 * np.pi * frequency * current_timestep + phase_shift)
        * np.ones_like(input_signal)
    )


def noise_disturbance(input_signal, mean, std):
    """Returns a random noise specified mean and a standard deviation.

    Args:
        input_signal (numpy.ndarray): The signal to which the disturbance should be
            applied. Used for determining the direction of the disturbance.
        mean

    Returns:
        numpy.ndarray: The disturbance array.
    """
    return np.random.normal(mean, std, len(input_signal),)
