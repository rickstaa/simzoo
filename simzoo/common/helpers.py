"""Functions that are used in multiple simzoo environments.
"""
color2num = dict(
    gray=30,
    red=31,
    green=32,
    yellow=33,
    blue=34,
    magenta=35,
    cyan=36,
    white=37,
    crimson=38,
)


def colorize(string, color, bold=False, highlight=False):
    """Colorize a string.

    This function was originally written by John Schulman.

    Args:
        string (str): The string you want to colorize.
        color (str): The color you want to use.
        bold (bool, optional): Whether you want the text to be bold text has to be bold.
        highlight (bool, optional):  Whether you want to highlight the text. Defaults to
            False.

    Returns:
        str: Colorized string.
    """
    if color:
        attr = []
        num = color2num[color]
        if highlight:
            num += 10
        attr.append(str(num))
        if bold:
            attr.append("1")
        return "\x1b[%sm%s\x1b[0m" % (";".join(attr), string)
    else:
        return string
