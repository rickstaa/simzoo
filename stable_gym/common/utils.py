"""Utility functions that are used in multiple Stable Gym gymnasium environments."""
import re

import numpy as np
from gymnasium.utils import colorize as gym_colorize


def colorize(string, color, bold=False, highlight=False):
    """Colorize a string.

    .. seealso::
        This function wraps the :meth:`gymnasium.utils.colorize` function to make sure
        that it also works with empty empty color strings.

    Args:
        string (str): The string you want to colorize.
        color (str): The color you want to use.
        bold (bool, optional): Whether you want the text to be bold text has to be bold.
        highlight (bool, optional):  Whether you want to highlight the text. Defaults to
            ``False``.

    Returns:
        str: Colorized string.
    """
    if color:  # If not empty.
        return gym_colorize(string, color, bold, highlight)
    else:
        return string


def get_flattened_values(input_obj):
    """Retrieves all the values that are present in a nested dictionary and appends them
    to a list. Its like a recursive version of the :meth:`dict.values()` method.

    Args:
        input_obj (dict): The input dictionary from which you want
            to retrieve all the values.

    Returns:
        list: A list containing all the values that were present in the nested
            dictionary.
    """
    flat_values = []
    if isinstance(input_obj, dict):
        for item in input_obj.values():
            if isinstance(item, dict):
                for it in item.values():
                    flat_values.extend(get_flattened_values(it))
            else:
                flat_values.append(item)
    else:
        flat_values.append(input_obj)
    return flat_values


def get_flattened_keys(input_obj, include_root=False):
    """Retrieves all the keys that are present in a nested dictionary and appends them
    to a list. Its like a recursive version of the :meth:`dict.keys()` method.

    Args:
        input_obj (dict): The input dictionary from which you want
            to retrieve all the keys.
        include_root (bool): Whether you want to include the root level keys. Defaults
            to ``False``.

    Returns:
        list: A list containing all the keys that were present in the nested
            dictionary.
    """
    flat_keys = []
    if isinstance(input_obj, dict):
        if include_root:
            flat_keys.extend(input_obj.keys())
        for key, val in input_obj.items():
            if isinstance(val, dict):
                flat_keys.extend(get_flattened_keys(val))
            else:
                flat_keys.append(key)
    else:
        flat_keys.append(input_obj)
    return flat_keys


def abbreviate(input_item, length=1, max_length=4, capitalize=True):
    """Creates unique abbreviations for a string or list of strings.

    Args:
        input_item (union[str, list]): The string of list of strings which you want to
            abbreviate.
        length (int, optional): The desired length of the abbreviation. Defaults to
            ``1``.
        max_length (int, optional): The maximum length of the abbreviation. Defaults to
            4.
        capitalize (bool, optional): Whether the abbrevaitions should be capitalized.
            Defaults to True.

    Returns:
        list: List with abbreviations.
    """
    if isinstance(input_item, list):
        items = []
        abbreviations = []
        for it in input_item:
            unique = False
            length_tmp = length
            suffix = ""
            while not unique:
                abbreviation = (
                    it[:length_tmp].capitalize() + str(suffix)
                    if capitalize
                    else it[:length_tmp] + suffix
                )
                if abbreviation not in abbreviations:  # Check if unique.
                    abbreviations.append(abbreviation)
                    items.append(it)
                    unique = True
                else:
                    prev_item = items[abbreviations.index(abbreviation)]
                    if it == prev_item:  # Allow if item was equal.
                        abbreviations.append(abbreviation)
                        items.append(it)
                        unique = True
                    else:  # Use longer abbreviation otherwise.
                        if length_tmp < max_length:
                            length_tmp += 1
                        else:
                            suffix = get_lowest_next_int(abbreviations)
        return abbreviations
    else:
        return input_item[:length].capitalize() if capitalize else input_item[:length]


def get_lowest_next_int(input_item):
    """Retrieves the lowest next integer that is not present in a string or float list.

    Args:
        input_item (union[int, str, list]): The input for which you want to determine
            the next lowest interger.

    Returns:
        int: The next lowest integer.
    """
    if isinstance(input_item, list):
        input_ints = [
            (
                round(float(re.sub("[^0-9.]", "", item)))
                if (re.sub("[^0-9.]", "", item) != "")
                else ""
            )
            if isinstance(item, str)
            else item
            for item in input_item
        ]  # Trim all non-numeric chars
        input_ints = [item for item in input_ints if item != ""]
    else:
        input_ints = [
            (
                (
                    round(float(re.sub("[^0-9.]", "", input_item)))
                    if (re.sub("[^0-9.]", "", input_item) != "")
                    else ""
                )
                if isinstance(input_item, str)
                else input_item
            )
        ]
    input_ints = input_ints if input_ints else [0]
    return list(set(input_ints) ^ set(range(min(input_ints), max(input_ints) + 2)))[0]


def friendly_list(input_list, apostrophes=False):
    """Transforms a list to a human friendly format (separated by commas and ampersand).

    Args:
        input_list (list): The input list.
        apostrophes(bool, optional): Whether the list items should be encapsuled with
            apostrophes. Defaults to ``False``.

    Returns:
        str: Human friendly list string.
    """
    input_list = (
        ["'" + item + "'" for item in input_list] if apostrophes else input_list
    )
    return " & ".join(", ".join(input_list).rsplit(", ", 1))


def strip_underscores(text, position="all"):
    """Strips leading and/or trailing underscores from a string.

    Args:
        text (str): The input string.
        position (str, optional): From which position underscores should be removed.
            Options are 'leading', 'trailing' & 'both'. Defaults to "both".

    Returns:
        str: String without the underscores.
    """
    if position.lower() == "leading":
        while text.startswith("_"):
            text = text[1:]
    elif position.lower() == "trailing":
        while text.endswith("_"):
            text = text[:-1]
    else:
        text = text.strip("_")
    return text


def inject_value(input_item, value, round_accuracy=2, order=False, axis=0):
    """Injects a value into a list or dictionary if it is not yet present.

    Args:
        input_item (union[list,dict]): The input list or dictionary.
        value (float): The value you want to inject.
        round_accuracy (int, optional): The accuracy used for checking whether a value
            is present. Defaults to 2.
        order (bool, optional): Whether the list should be ordered when returned.
            Defaults to ``false``.
        axis (int, optional): The axis along which you want to inject the value. Only
            used when the input is a numpy array. Defaults to ``0``.

    Returns:
        union[list,dict]: The list or dictionary that contains the value.
    """
    order_op = (
        lambda *args, **kwargs: sorted(*args, **kwargs)
        if order
        else list(*args, **kwargs)
    )
    if isinstance(input_item, dict):
        return {
            k: inject_value(
                v, value=value, round_accuracy=round_accuracy, order=order, axis=axis
            )
            for k, v in input_item.items()
        }
    elif isinstance(input_item, np.ndarray) and input_item.ndim > 1:
        transpose_matrix = np.eye(input_item.ndim, dtype=np.int16)
        return np.transpose(
            np.array(
                [
                    order_op([value] + [it for it in item if it != value])
                    for item in np.transpose(input_item, transpose_matrix[axis])
                ]
            ),
            transpose_matrix[axis],
        )
    else:
        return order_op([value] + [item for item in input_item if item != value])


def verify_number_and_cast(x):
    """Verify parameter is a single number and cast to a float."""
    try:
        x = float(x)
    except (ValueError, TypeError) as e:
        raise ValueError(f"An option ({x}) could not be converted to a float.") from e
    return x


def maybe_parse_reset_bounds(options, default_low, default_high):
    """sThis function can be called during a reset() to customize the sampling
    ranges for setting the initial state distributions.

    Args:
      options: Options passed in to reset().
      default_low: Default lower limit to use, if none specified in options.
      default_high: Default upper limit to use, if none specified in options.

    Returns:
        (tuple): a tuple containing:

            - low (:obj:`np.ndarray`): Lower limit for each dimension.
            - high ():obj:`np.ndarray`): Upper limit for each dimension.
    """
    if options is None:
        return default_low, default_high

    low = options.get("low") if "low" in options else default_low
    high = options.get("high") if "high" in options else default_high

    # We expect only numerical inputs.
    low = verify_number_and_cast(low)
    high = verify_number_and_cast(high)
    if low > high:
        raise ValueError(
            f"Lower bound ({low}) must be lower than higher bound ({high})."
        )

    return low, high
