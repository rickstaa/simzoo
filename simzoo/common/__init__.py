"""Contains several functions that can be used across all the simzoo environments.
"""
from simzoo.common.disturber import Disturber  # noqa: F401
