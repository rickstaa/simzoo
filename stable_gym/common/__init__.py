"""Contains several functions that can be used across all the Stable Gym gymnasium
environments.
"""
from stable_gym.common.disturber import Disturber  # noqa: F401
