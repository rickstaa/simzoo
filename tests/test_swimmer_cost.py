"""Test if the SwimmerCost environment still behaves like the original Swimmer
environment when the same environment parameters are used.
"""
import gymnasium as gym
import numpy as np
from gymnasium.logger import ERROR

from stable_gym.common.utils import change_precision

gym.logger.set_level(ERROR)

PRECISION = 16


class TestSwimmerCostEqual:
    # Make original Swimmer environment.
    env = gym.make("Swimmer")
    # Make SwimmerCost environment.
    env_cost = gym.make(
        "SwimmerCost",
        exclude_reference_from_observation=True,
        exclude_x_velocity_from_observation=True,
    )

    def test_equal_reset(self):
        """Test if reset behaves the same."""
        # Perform reset and check if observations are equal.
        observation, _ = self.env.reset(seed=42)
        observation_cost, _ = self.env_cost.reset(seed=42)
        assert np.allclose(
            observation, observation_cost
        ), f"{observation} != {observation_cost}"

    def test_equal_steps(self):
        """Test if steps behave the same."""
        # Perform several steps and check if observations are equal.
        self.env.reset(seed=42), self.env_cost.reset(seed=42)
        for _ in range(10):
            self.env.action_space.seed(42)
            action = self.env.action_space.sample()
            observation, _, _, _, _ = self.env.step(action)
            observation_cost, _, _, _, _ = self.env_cost.step(action)
            assert np.allclose(
                observation, observation_cost
            ), f"{observation} != {observation_cost}"

    # NOTE: We decrease the test precision to 16 decimals to ignore numerical
    # differences due to hardware or library differences.
    def test_snapshot(self, snapshot):
        """Test if the 'SwimmerCost' environment is still equal to snapshot."""
        self.env_cost = gym.make(
            "SwimmerCost", exclude_reference_error_from_observation=False
        )  # Check full observation.
        observation, info = self.env_cost.reset(seed=42)
        assert (change_precision(observation, precision=PRECISION) == snapshot).all()
        assert change_precision(info, precision=PRECISION) == snapshot
        self.env_cost.action_space.seed(42)
        for _ in range(5):
            action = self.env_cost.action_space.sample()
            observation, reward, terminated, truncated, info = self.env_cost.step(
                action
            )
            assert (
                change_precision(observation, precision=PRECISION) == snapshot
            ).all()
            assert change_precision(reward, precision=PRECISION) == snapshot
            assert terminated == snapshot
            assert truncated == snapshot
            assert change_precision(info, precision=PRECISION) == snapshot
