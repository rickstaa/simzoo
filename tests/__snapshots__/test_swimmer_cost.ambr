# serializer version: 1
# name: TestSwimmerCostEqual.test_snapshot
  0.0717195839822765
# ---
# name: TestSwimmerCostEqual.test_snapshot.1
  0.0394736058118728
# ---
# name: TestSwimmerCostEqual.test_snapshot.10
  0.0
# ---
# name: TestSwimmerCostEqual.test_snapshot.11
  dict({
  })
# ---
# name: TestSwimmerCostEqual.test_snapshot.12
  0.0586930618000772
# ---
# name: TestSwimmerCostEqual.test_snapshot.13
  0.0649009788475318
# ---
# name: TestSwimmerCostEqual.test_snapshot.14
  -0.1030793695707376
# ---
# name: TestSwimmerCostEqual.test_snapshot.15
  0.026767619662024
# ---
# name: TestSwimmerCostEqual.test_snapshot.16
  0.7977753109453727
# ---
# name: TestSwimmerCostEqual.test_snapshot.17
  -0.6946564587215142
# ---
# name: TestSwimmerCostEqual.test_snapshot.18
  1.3203403932370403
# ---
# name: TestSwimmerCostEqual.test_snapshot.19
  -1.0667011828424493
# ---
# name: TestSwimmerCostEqual.test_snapshot.2
  -0.0811645304224701
# ---
# name: TestSwimmerCostEqual.test_snapshot.20
  1.0
# ---
# name: TestSwimmerCostEqual.test_snapshot.21
  -0.940159608181165
# ---
# name: TestSwimmerCostEqual.test_snapshot.22
  0.0598403918188351
# ---
# name: TestSwimmerCostEqual.test_snapshot.23
  0.8839000888553616
# ---
# name: TestSwimmerCostEqual.test_snapshot.24
  False
# ---
# name: TestSwimmerCostEqual.test_snapshot.25
  False
# ---
# name: TestSwimmerCostEqual.test_snapshot.26
  dict({
    'cost_ctrl': 3.1515109539e-05,
    'cost_velocity': 0.8839000888553616,
    'distance_from_origin': 0.0574074063134864,
    'x_position': 0.0571848253839461,
    'x_velocity': 0.0598403918188351,
    'y_position': 0.0050503510223858,
    'y_velocity': 0.4318665767993828,
  })
# ---
# name: TestSwimmerCostEqual.test_snapshot.27
  0.0184752597769699
# ---
# name: TestSwimmerCostEqual.test_snapshot.28
  0.1346491521755371
# ---
# name: TestSwimmerCostEqual.test_snapshot.29
  -0.1457052992158369
# ---
# name: TestSwimmerCostEqual.test_snapshot.3
  0.0951244703273512
# ---
# name: TestSwimmerCostEqual.test_snapshot.30
  0.0116081705231652
# ---
# name: TestSwimmerCostEqual.test_snapshot.31
  1.3275948136464797
# ---
# name: TestSwimmerCostEqual.test_snapshot.32
  -1.2872296796028666
# ---
# name: TestSwimmerCostEqual.test_snapshot.33
  2.124422100938903
# ---
# name: TestSwimmerCostEqual.test_snapshot.34
  -1.050781983705994
# ---
# name: TestSwimmerCostEqual.test_snapshot.35
  1.0
# ---
# name: TestSwimmerCostEqual.test_snapshot.36
  -0.9850483900707502
# ---
# name: TestSwimmerCostEqual.test_snapshot.37
  0.0149516099292497
# ---
# name: TestSwimmerCostEqual.test_snapshot.38
  0.970320330780977
# ---
# name: TestSwimmerCostEqual.test_snapshot.39
  False
# ---
# name: TestSwimmerCostEqual.test_snapshot.4
  0.0522279403980706
# ---
# name: TestSwimmerCostEqual.test_snapshot.40
  False
# ---
# name: TestSwimmerCostEqual.test_snapshot.41
  dict({
    'cost_ctrl': 6.70186460018e-05,
    'cost_velocity': 0.970320330780977,
    'distance_from_origin': 0.0751820903036745,
    'x_position': 0.0577828897811161,
    'x_velocity': 0.0149516099292497,
    'y_position': 0.048098693859327,
    'y_velocity': 1.0762085709235305,
  })
# ---
# name: TestSwimmerCostEqual.test_snapshot.42
  -0.0013654562937075
# ---
# name: TestSwimmerCostEqual.test_snapshot.43
  0.1529612332667866
# ---
# name: TestSwimmerCostEqual.test_snapshot.44
  -0.1206629739894392
# ---
# name: TestSwimmerCostEqual.test_snapshot.45
  0.150390800293635
# ---
# name: TestSwimmerCostEqual.test_snapshot.46
  -0.2945187732354072
# ---
# name: TestSwimmerCostEqual.test_snapshot.47
  0.2554889787617607
# ---
# name: TestSwimmerCostEqual.test_snapshot.48
  -1.1301376727439139
# ---
# name: TestSwimmerCostEqual.test_snapshot.49
  2.220792926335492
# ---
# name: TestSwimmerCostEqual.test_snapshot.5
  0.0572128610553908
# ---
# name: TestSwimmerCostEqual.test_snapshot.50
  1.0
# ---
# name: TestSwimmerCostEqual.test_snapshot.51
  -0.910529719768756
# ---
# name: TestSwimmerCostEqual.test_snapshot.52
  0.0894702802312439
# ---
# name: TestSwimmerCostEqual.test_snapshot.53
  0.8290643705821694
# ---
# name: TestSwimmerCostEqual.test_snapshot.54
  False
# ---
# name: TestSwimmerCostEqual.test_snapshot.55
  False
# ---
# name: TestSwimmerCostEqual.test_snapshot.56
  dict({
    'cost_ctrl': 0.0001563634634018,
    'cost_velocity': 0.8290643705821694,
    'distance_from_origin': 0.0915890070175345,
    'x_position': 0.0613617009903658,
    'x_velocity': 0.0894702802312439,
    'y_position': 0.067994763460335,
    'y_velocity': 0.4974017400252014,
  })
# ---
# name: TestSwimmerCostEqual.test_snapshot.57
  0.0003841448162947
# ---
# name: TestSwimmerCostEqual.test_snapshot.58
  0.11788280333712
# ---
# name: TestSwimmerCostEqual.test_snapshot.59
  -0.0288204456102053
# ---
# name: TestSwimmerCostEqual.test_snapshot.6
  -0.0743772734648908
# ---
# name: TestSwimmerCostEqual.test_snapshot.60
  0.1155874118811639
# ---
# name: TestSwimmerCostEqual.test_snapshot.61
  0.0486519142291015
# ---
# name: TestSwimmerCostEqual.test_snapshot.62
  -0.1630316639903109
# ---
# name: TestSwimmerCostEqual.test_snapshot.63
  -0.624124384806579
# ---
# name: TestSwimmerCostEqual.test_snapshot.64
  2.350761838731597
# ---
# name: TestSwimmerCostEqual.test_snapshot.65
  1.0
# ---
# name: TestSwimmerCostEqual.test_snapshot.66
  -0.8667111829722601
# ---
# name: TestSwimmerCostEqual.test_snapshot.67
  0.1332888170277399
# ---
# name: TestSwimmerCostEqual.test_snapshot.68
  0.7511882746891746
# ---
# name: TestSwimmerCostEqual.test_snapshot.69
  False
# ---
# name: TestSwimmerCostEqual.test_snapshot.7
  -0.0099228124208866
# ---
# name: TestSwimmerCostEqual.test_snapshot.70
  False
# ---
# name: TestSwimmerCostEqual.test_snapshot.71
  dict({
    'cost_ctrl': 6.00106835365e-05,
    'cost_velocity': 0.7511882746891746,
    'distance_from_origin': 0.0918631916858829,
    'x_position': 0.0666932536714754,
    'x_velocity': 0.1332888170277399,
    'y_position': 0.0631732213950618,
    'y_velocity': -0.1205385516318314,
  })
# ---
# name: TestSwimmerCostEqual.test_snapshot.72
  0.0117915594668683
# ---
# name: TestSwimmerCostEqual.test_snapshot.73
  0.0658718157028656
# ---
# name: TestSwimmerCostEqual.test_snapshot.74
  0.0721597648275078
# ---
# name: TestSwimmerCostEqual.test_snapshot.75
  0.1098815986431019
# ---
# name: TestSwimmerCostEqual.test_snapshot.76
  -0.7852810350118694
# ---
# name: TestSwimmerCostEqual.test_snapshot.77
  0.712476930243054
# ---
# name: TestSwimmerCostEqual.test_snapshot.78
  -1.9431447627877871
# ---
# name: TestSwimmerCostEqual.test_snapshot.79
  2.6880029820455054
# ---
# name: TestSwimmerCostEqual.test_snapshot.8
  1.0
# ---
# name: TestSwimmerCostEqual.test_snapshot.80
  1.0
# ---
# name: TestSwimmerCostEqual.test_snapshot.81
  -0.8864606174497422
# ---
# name: TestSwimmerCostEqual.test_snapshot.82
  0.1135393825502578
# ---
# name: TestSwimmerCostEqual.test_snapshot.83
  0.7858124262893783
# ---
# name: TestSwimmerCostEqual.test_snapshot.84
  False
# ---
# name: TestSwimmerCostEqual.test_snapshot.85
  False
# ---
# name: TestSwimmerCostEqual.test_snapshot.86
  dict({
    'cost_ctrl': 5.63044130802e-05,
    'cost_velocity': 0.7858124262893783,
    'distance_from_origin': 0.0858991719326769,
    'x_position': 0.0712348289734857,
    'x_velocity': 0.1135393825502578,
    'y_position': 0.0480027799178113,
    'y_velocity': -0.3792610369312622,
  })
# ---
# name: TestSwimmerCostEqual.test_snapshot.9
  -1.0
# ---
